src/net/server/bytearrstream.rs

```diff
@@ -0,0 +1,452 @@
+use crate::{error::BoxError, ff, net::MpcHelperServerError};
+use async_trait::async_trait;
+use axum::extract::{BodyStream, FromRequest, Query, RequestParts};
+use futures::{ready, Stream};
+use hyper::body::{Bytes, HttpBody};
+use pin_project::pin_project;
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// the size of the field type specified in the query param of the request
+/// Possible sizes are are currently from:
+/// * `fp2`
+/// * `fp31`
+/// * `fp32_bit_prime`
+struct FieldSize {
+    field_size: u32,
+}
+
+#[cfg(feature = "enable-serde")]
+#[async_trait]
+impl<B: Send> FromRequest<B> for FieldSize {
+    type Rejection = MpcHelperServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        #[derive(serde::Deserialize)]
+        struct FieldStr {
+            field_type: String,
+        }
+
+        let Query(FieldStr { field_type }) = req.extract().await?;
+        let field_size = ff::size_in_bytes_from_type_str(&field_type)?;
+
+        Ok(Self { field_size })
+    }
+}
+
+/// Wraps a [`BodyStream`] and produce a new stream that has chunks of exactly size `size_in_bytes`.
+/// # Errors
+/// If the downstream body is not a multiple of `size_in_bytes`.
+#[derive(Debug)]
+#[pin_project]
+pub struct ByteArrStream {
+    #[pin]
+    body: BodyStream,
+    size_in_bytes: u32,
+    buffered_size: u32,
+    buffered: VecDeque<Bytes>,
+}
+
+impl ByteArrStream {
+    /// # Panics
+    /// if `size_in_bytes` is 0
+    pub fn new(body: BodyStream, size_in_bytes: u32) -> Self {
+        assert_ne!(size_in_bytes, 0);
+        Self {
+            body,
+            size_in_bytes,
+            buffered_size: 0,
+            buffered: VecDeque::new(),
+        }
+    }
+
+    /// returns [`Bytes`] of `size_in_bytes` length from the buffered chunks. Returns [`None`] if
+    /// there are less than `size_in_bytes` bytes in the buffer
+    fn pop_front(&mut self) -> Option<Bytes> {
+        // not enough bytes buffered
+        if self.buffered_size < self.size_in_bytes {
+            None
+        } else if self.buffered[0].len() > self.size_in_bytes as usize {
+            // if the first buffer is large enough, simply split it in-place.
+            // This is O(1) operation, and should be true majority of the time, assuming large size
+            // buffers
+            let out_bytes = self.buffered[0].split_to(self.size_in_bytes as usize);
+            self.buffered_size -= self.size_in_bytes;
+            Some(out_bytes)
+        } else {
+            // First buffer is too small, so we need to take across buffers.
+            // Will require a memcopy of u8's across buffers in order to create 1 contiguous buffer
+            // to return.
+            // Removes expended buffers after taking
+
+            let mut out_bytes = Vec::with_capacity(self.size_in_bytes as usize);
```

Comment:
Please avoid allocating on the heap for this.  It is OK to allocate a fixed 8 bytes into an array for this.

Reply:
the problem is that I don't know what `size_in_bytes` is at compile time. are you proposing we have a max of `8`?

Reply:
`BitArr` can do it at runtime I think

Reply:
i took a look, and it seems like it's saying lengths must be const: https://docs.rs/bitvec/0.22.3/bitvec/macro.BitArr.html

Reply:
but field size is const, isn't it?